### DISCUSSION TITLE
How can I get a function to run on every loop of axum? (Twilio-rs integration)

### DISCUSSION THREAD

**Discussion by matthiasdebernardini** [tokio-rs/axum#2002] [Ideas]
I'd like to receive an SMS when my server is down but not run an extra piece of software (some python or shell script), I'd like it to.

There is a crate that implements this so that I only have to run a little bit of configuation code once, then run a function on every nth loop that axum runs. Ideally to call twilio API, like [so](https://github.com/neil-lobracco/twilio-rs/blob/master/src/lib.rs). 

Where would I put this? 

I'd actually like this to be part of the axum builder pattern by implementing the tower::Service() trait, is there an example of someone integrating something similar to this? This way I can publish a crate and other's can easily call it.

**Comment by davidpdrsn** [comment]
What do you mean by "on every nth loop that axum runs". Do you mean run something for every request? If so you need a middleware. See `axum::middleware::from_fn`

**Reply by matthiasdebernardini** [reply]
axum is a framework that implements a web server, to implement a server you have to have a while loop somewhere, don't you?

```rust
loop {
    if there_is_a_reason_to_stop {
        panic!()
    } else {
        keep_running()
    }
}
```

so when that loop runs (if it does run at all - if this is not the case then I need to study axum better) I'd like it to check additional conditions as well

**Reply by davidpdrsn** [reply]
You mean [graceful shutdown](https://github.com/tokio-rs/axum/blob/main/examples/graceful-shutdown/src/main.rs)?

**Comment by matthiasdebernardini** [comment]
Actually, I think this is what I would need to implement. https://docs.rs/hyper/0.14.26/hyper/server/struct.Builder.html

**Reply by davidpdrsn** [reply]
Are you saying you want to implement your own server builder? Not sure what you mean 😅

**Reply by matthiasdebernardini** [reply]
My apologies, 

I want to configure and add this functionality using the builder pattern, like so;

```rust 
    let twilio = configure(params);

    Server::bind(&addr)
        .add_extra_thing(twilio)
        .serve(service)
        .await
        .expect("server failed");
```

**Reply by davidpdrsn** [reply]
What would `.add_extra_thing(twilio)` do?

**Reply by davidpdrsn** [reply]
You can run other async things concurrently using [`tokio::spawn`](https://docs.rs/tokio/latest/tokio/task/fn.spawn.html) if that's what you want.

**Reply by matthiasdebernardini** [reply]
`add_extra_thing(twilio)` would create an http Client struct, connect to the API and make an http get request periodically, at some time interval in my case.

That makes sense to use that, thanks.
